providers/netty4/src/main/java/org/asynchttpclient/netty/request/body/NettyReactiveStreamBody.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2010-2012 Sonatype, Inc. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package org.asynchttpclient.netty.request.body;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.NoSuchElementException;
+
+import org.asynchttpclient.netty.NettyResponseFuture;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.typesafe.netty.HandlerSubscriber;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.handler.codec.http.DefaultHttpContent;
+import io.netty.handler.codec.http.HttpContent;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.util.concurrent.EventExecutor;
+
+public class NettyReactiveStreamBody implements NettyBody {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyReactiveStreamBody.class);
+    private static final String NAME_IN_CHANNEL_PIPELINE = "request-body-streamer";
+
+    private final Publisher<ByteBuffer> publisher;
+
+    public NettyReactiveStreamBody(Publisher<ByteBuffer> publisher) {
+        this.publisher = publisher;
+    }
+
+    @Override
+    public long getContentLength() {
+        return -1L;
+    }
+
+    @Override
+    public String getContentType() {
+        return null;
+    }
+
+    @Override
+    public void write(Channel channel, NettyResponseFuture<?> future) throws IOException {
+        if (future.isStreamWasAlreadyConsumed()) {
+            LOGGER.warn("Stream has already been consumed and cannot be reset");
+        } else {
+            future.setStreamWasAlreadyConsumed(true);
+            NettySubscriber subscriber = new NettySubscriber(channel);
+            channel.pipeline().addLast(NAME_IN_CHANNEL_PIPELINE, subscriber);
+            publisher.subscribe(new SubscriberAdapter(subscriber));
+        }
+    }
+
+    private static class SubscriberAdapter implements Subscriber<ByteBuffer> {
```

Comment:
A better "Netty" way to implement this would be to insert another handler in the chain, just before the `HandlerSubscriber` that intercepts calls to `write`, and converts it to a `DefaultHttpContent`.